osc-server/src/main/java/org/osc/core/broker/service/validator/DeleteDistributedApplianceRequestValidator.java

```diff
@@ -53,13 +59,46 @@ public DistributedAppliance validateAndLoad(BaseDeleteRequest request) throws Ex
                             da.getId()));
         }
 
+        validateDAReferenceToSfcAndSecurityGroup(em, da);
+
         if (!da.getMarkedForDeletion() && request.isForceDelete()) {
             throw new VmidcBrokerValidationException(
                     "Distributed Appilance with ID "
                             + request.getId()
                             + " is not marked for deletion and force delete operation is applicable only for entries marked for deletion.");
         }
-
         return da;
     }
+
+    private void validateDAReferenceToSfcAndSecurityGroup(EntityManager em, DistributedAppliance da) throws Exception {
+        for (VirtualSystem vs : da.getVirtualSystems()) {
+            try {
+                validateVSReferenceToSfcAndSecurityGroup(em, vs);
+            } catch (VmidcBrokerValidationException ex) {
+                throw new VmidcBrokerValidationException(String.format("Cannot delete the distributed appliance '%s' reason:",da.getName())+" "+ex.getMessage());
+            }
+        }
+    }
+
+    private void validateVSReferenceToSfcAndSecurityGroup(EntityManager em, VirtualSystem vs) throws Exception {
+        if (vs.getServiceFunctionChains() == null) {
+            return;
+        }
+        for (ServiceFunctionChain sfc : vs.getServiceFunctionChains()) {
+            validateSfcReferenceToSecurityGroup(em, sfc);
+        }
+    }
+
+    private void validateSfcReferenceToSecurityGroup(EntityManager em, ServiceFunctionChain sfc) throws Exception {
+        List<SecurityGroup> sgList = SecurityGroupEntityMgr.listSecurityGroupsBySfcId(em, sfc.getId());
+        String sgNames = sgList.stream().filter(sg -> !sg.getMarkedForDeletion()).map(sg -> sg.getName())
+                .collect(Collectors.joining(", "));
+        if (!sgNames.isEmpty()) {
+            throw new VmidcBrokerValidationException(String.format("Service Function Chain '%s' and binded to a Security Group(s) '%s'",
+                    sfc.getName(), sgNames));
+        } else if (!sfc.getMarkedForDeletion() && sfc != null) {
+            throw new VmidcBrokerValidationException(String.format("Service Function Chain '%s'",
```

Comment:
This msg does not make sense.... Service Function Chain ?? The message must be a complete statement.